db/schema_tables.cc

```diff
@@ -1668,6 +1706,48 @@ static future<> merge_aggregates(distributed<service::storage_proxy>& proxy, sch
     co_await merge_functions(proxy, before, after, create_aggregate);
 }
 
+static future<> merge_scylla_aggregates(distributed<service::storage_proxy>& proxy, schema_result before, schema_result after) {
+    auto diff = diff_rows(before, after);
+
+    auto update_reducefunc = [&] (replica::database& db, const query::result_set_row& row) {
+        auto ks = row.get_nonnull<sstring>("keyspace_name");
+        auto a_name = row.get_nonnull<sstring>("aggregate_name");
+        auto r_name = row.get_nonnull<sstring>("reduce_func");
+
+        auto types = read_arg_types(db, row, ks);
+        data_type state_type = db::cql_type_parser::parse(ks, row.get_nonnull<sstring>("state_type"), db.user_types());
+
+
+        auto aggr = dynamic_pointer_cast<cql3::functions::user_aggregate>(
+            cql3::functions::functions::find(cql3::functions::function_name{ks, a_name}, types)
+        );
+        if (!aggr) {
+            throw std::runtime_error(format("User-defined aggragate {} not found.", a_name));
+        }
+
+        auto reducer = dynamic_pointer_cast<cql3::functions::scalar_function>(
+            cql3::functions::functions::find(cql3::functions::function_name{ks, r_name}, {state_type, state_type})
+        );
+        if (!reducer) {
+            throw std::runtime_error(format("Scalar function {} used as reduce function not found.", r_name));
+        }
+
+        aggr->set_reducefunc(std::move(reducer));
+    };
+
+
+    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) {
+        for (const auto& val : diff.created) {
+            update_reducefunc(db, *val);
+        }
+        for (const auto& val : diff.altered) {
+            update_reducefunc(db, *val);
```

Comment:
Isn't this creating a window where the reduce function is bad? After merge_functions() is called but before mesge_scylla_aggregates() is called.

If so we should fold merge_scylla_aggregates into merge_functions so it happens simultaneously. This will allow us to drop the mutating set_reducefunc.

Reply:
Merge merge_scylla_aggregates into merge_aggregates().

Nor sure what you meant be "really simultaneously".

Reply:
I've joined the diff from both tables, however I've ignored the altered rows for now. (more detailed comment in changelog)